samza-api/src/main/java/org/apache/samza/operators/windows/Windows.java

```diff
@@ -16,85 +16,378 @@
+   * {@link org.apache.samza.operators.triggers.Triggers} APIs.
+   *
+   * For example, to window a stream into count based windows that trigger every 50 messages or every 10 minutes.
+   * <pre> {@code
+   * MessageStream<Long> stream = ...;
+   * MessageStream<WindowOutput<WindowKey, Collection<Long>>> windowed = stream.window(Windows.globalWindow()
+   * .setEarlyTrigger(Triggers.repeat(Triggers.any(Triggers.count(50), Triggers.timeSinceFirstMessage(Duration.seconds(10))))))
+   * }
+   * </pre>
    *
-   * @param sessionKeyFunction  function to calculate session window key
-   * @param <M>  type of the input {@link MessageEnvelope}
-   * @param <WK>  type of the session window key
-   * @return  the {@link Window} function for the session
+   * @param <M> the type of {@link MessageEnvelope}
+   * @return the created {@link WindowFunction} function.
    */
-  public static <M extends MessageEnvelope, WK> Window<M, WK, Integer, WindowOutput<WK, Integer>> intoSessionCounter(Function<M, WK> sessionKeyFunction) {
-    return new SessionWindow<>(sessionKeyFunction, (m, c) -> c + 1);
+
+  public static <M extends MessageEnvelope> WindowFunction<M, Void, WindowKey<Void>, Collection<M>, WindowOutput<WindowKey<Void>, Collection<M>>> globalWindow() {
+    BiFunction<M, Collection<M>, Collection<M>> aggregator = (m, c) -> {
+      c.add(m);
+      return c;
+    };
+    return globalWindow(aggregator);
   }
 
+  /**
+   * Returns a global {@link WindowFunction} that windows its values per-key. The window does not have a default trigger.
+   * The triggering behavior must be specified by setting early
+   * triggers using the {@link org.apache.samza.operators.triggers.Triggers} APIs.
+   *
+   * <p> The below example windows the stream into count based windows. The window triggers every 50 messages or every
+   * 10 minutes.
+   *
+   * <pre> {@code
+   * MessageStream<UserClick> stream = ...;
+   * BiFunction<UserClick, Long, Long> maxAggregator = (m, c)-> Math.max(parseLongField(m), c);
+   * Function<UserClick, String> keyFn = ...;
+   * MessageStream<WindowOutput<WindowKey<String>, Long>> windowed = stream.window(Windows.keyedGlobalWindow(keyFn, maxAggregator)
+   * .setEarlyTrigger(Triggers.repeat(Triggers.any(Triggers.count(50), Triggers.timeSinceFirstMessage(Duration.minutes(10))))))
+   * }
+   * </pre>
+   *
+   * @param keyFn the function to extract the key from a {@link MessageEnvelope}
+   * @param aggregateFunction the function to aggregate window results
+   * @param <M> the type of {@link MessageEnvelope}
+   * @param <K> type of the key in the {@link WindowFunction}
+   * @param <WV> type of the {@link WindowOutput} output value
+   * @return the created {@link WindowFunction} function
+   */
+
+  public static <M extends MessageEnvelope, K, WV> WindowFunction<M, K, WindowKey<K>, WV, WindowOutput<WindowKey<K>, WV>> keyedGlobalWindow(Function<M, K> keyFn, BiFunction<M, WV, WV> aggregateFunction) {
```

Comment:
>> I think fundamentally I have a confusion here on the proposed global window. 
Here are some properties of a `globalWindow`:
1. The proposed `globalWindow` is an operator that does not contain any default trigger. 
2. The `globalWindow` has only a single pane that fires based on arbitrary triggering criteria. 
3. `globalWindow`s can be grouped by key (in which case, the triggering behavior is per-group-by key).

>> if the count is used as aggregate function, the global window will output (null, 3), (null, 5), (null, 9).
Nope, the output will be (WindowKey1, 3), (WK2, 5), (WK3, 9).
Each `WK` key object will have different metadata (start time, end time etc.). The `sink` function can choose what key it wants to use when it externalizes it. 

> Can we discuss more on this one?
Absolutely, happy to discuss this in-person.